### PR TITLE
llvm 6: init

### PR DESCRIPTION
**Pull request by dtzWill**
###### Motivation for this change

Start building and testing LLVM 6, currently on RC1.

Helps ensure release doesn't have bugs re:NixOS / our projects, and hopefully enables us to at least make LLVM 6 available after final release and update the default shortly after.

Anyway, test test please! :)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
I introduced a failure in the evaluator today, sorry about that. Testing for a fix:

@GrahamcOfBorg eval

**Comment by dtzWill** [comment]
@GrahamcOfBorg build clang_6 lld_6 lldb_6

File: pkgs/development/compilers/llvm/6/llvm.nix [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+  unpackPhase = ''
+    unpackFile ${src}
+    mv llvm-${version}* llvm
+    sourceRoot=$PWD/llvm
+    unpackFile ${compiler-rt_src}
+    mv compiler-rt-* $sourceRoot/projects/compiler-rt
+  '';
+
+  outputs = [ "out" ]
```
Comment (LnL7) [inline pkgs/development/compilers/llvm/6/llvm.nix]: Does this need something similar to https://github.com/NixOS/nixpkgs/pull/33871 to get python out of the closure?

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -6314,13 +6315,16 @@ with pkgs;
...
+  lld_6 = llvmPackages_6.lld;
 
   lldb = llvmPackages.lldb;
   lldb_4 = llvmPackages_4.lldb;
   lldb_5 = llvmPackages_5.lldb;
+  lldb_6 = llvmPackages_6.lldb;
 
   llvm = llvmPackages.llvm;
 
+  llvm_6  = llvmPackages_6.llvm;
```
Comment (LnL7) [inline pkgs/top-level/all-packages.nix]: Should we keep exposing all of these top-level?

File: pkgs/development/compilers/llvm/6/llvm.nix [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+let
+  src = fetch "llvm" "0dsik1l6jshd617fh4qs1373s6dfkq4gsjhyf6zq2qb6kyisy2vd";
+
+  # Used when creating a version-suffixed symlink of libLLVM.dylib
+  shortVersion = with stdenv.lib;
+    concatStringsSep "." (take 2 (splitString "." release_version));
+in stdenv.mkDerivation (rec {
+  name = "llvm-${version}";
+
+  unpackPhase = ''
```
Comment (LnL7) [inline pkgs/development/compilers/llvm/6/llvm.nix]: I cleaned this up on a test branch recently, let me see if I can find it.

**Review by LnL7 — COMMENTED** [review]
Everything builds on darwin.

File: pkgs/development/compilers/llvm/6/llvm.nix (hunk 1 above)
Comment (dtzWill) [reply pkgs/development/compilers/llvm/6/llvm.nix]: Good call.  Added.

**Comment by dtzWill** [comment]
rc2 has been tagged, but appears they don't upload the source tarballs until the "release testers" all build and upload binaries, so should be available in a few days.   Not worth bothering with grabbing off of SVN in the meantime, IMO.

**Comment by LnL7** [comment]
I don't think this is urgent since the build works.

**Comment by dtzWill** [comment]
rebased and fixed merge conflicts.  Added testing variants like we have for other versions.

Will be adding musl compat fixes too, now that they're in master as well :+1:.

**Comment by dtzWill** [comment]
Postponing musl fixes here while resolving discussion about musl ...

**Comment by dtzWill** [comment]
rebased, rc2 -> rc3 \o/

**Comment by dtzWill** [comment]
Hold on to your lugnuts! LLVM 6 final has been tagged:

http://lists.llvm.org/pipermail/release-testers/2018-March/000667.html

Will be a few more days for them to prepare the release and official source tarballs but .... soon!!

**Comment by LnL7** [comment]
Great!

**Comment by dtzWill** [comment]
Release notes: http://releases.llvm.org/6.0.0/docs/ReleaseNotes.html

Note this adds support for "retpolines", which may be of interest to those using clang-based stdenv's.

**Comment by data-man** [comment]
Sorry, but how do I use the clang 6?
I just started using the NixOS.

**Comment by dtzWill** [comment]
I don't think it's hit any channels yet, so in the meantime you can use something like this:

```console
$ NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/master.tar.gz nix-shell -p clang_6
```

and use `clang` within that shell.  Note that "installing" clang generally doesn't work as you may expect, there's an issue discussing this.

**Pull request merged**